src/Symfony/Bridge/Doctrine/Security/MagicLink/MagicLoginLinkTokenEntityTrait.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\Security\MagicLink;
+
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * Trait that can be used by your entity to help implement StoredMagicLinkTokenInterface.
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+trait MagicLoginLinkTokenEntityTrait
+{
+    /**
+     * @ORM\Column(type="string", length=20)
+     */
+    private $selector;
+
+    /**
+     * @ORM\Column(type="string", length=100)
+     */
+    private $hashedVerifier;
+
+    /**
+     * @ORM\Column(type="datetime_immutable")
```

Comment:
I would use `datetimetz_immutable`. On MySQL, this makes no difference (it does not support timezones so this is stored as a `datetime` field in the DB too). But other platforms (postgresql for instance) support it, which avoids relying on the default timezone of the database server.

Reply:
Just asking ... shouldn't we store all these dates in UTC and forget timezones precisely to avoid misconfiguration issues?

Reply:
Would converting to UTC work? The flow would be:

A) My `php.ini` uses America/New_York, which is UTC -4 at the moment.
B) We create a DateTime for right now, which is `2020-09-15 12:00:00`.
C) We convert that to UTC, so the internal string in the date is now `2020-09-15 16:00:00`
D) That UTC date string (e.g.`2020-09-15 16:00:00) is stored as a string in the database (even if we use `datetimetz_immutable`, that wouldn't store any timezone information for MySQL)
E) We query for this entity, and the string `2020-09-15 16:00:00` is converted into a DateTime object using `America/New_York`. And so, it now appears to be 4 hours *later* than it actually is.

Is it possible at all - reliably - to store the "expiresAt" with a datetime? Or must we instead use an integer timestamp?

Reply:
We're using [UTC converters](https://github.com/Drenso/symfony-shared/blob/master/src/Database/Types/UTCDateTimeType.php) for all datetime fields in our databases: this ensures that it is always stored as UTC (so PHP converts it if the `DateTime` object isn't), and always retrieved as UTC (by parsing the string directly).

However, I've only tested this on MySQL/MariaDB for now, so I cannot comment on other databases at this time.

Reply:
Wait, but if we just leave the date alone, then the database doesn’t affect this in any way! My date would start as America/New_York, get stored in the database with the time for this timezone, then doctrine would re-create the DateTime with this timezone.

So the true solution is to do absolutely nothing, isn’t it? Leave the date in whatever timezone it started in and that’s that. We could use the datetimetz as stof suggested (and we should) but that’s just a “nice to have” so that it’s more obvious what timezone is being used if you look in the database. 

Right?

Reply:
In the case of the date only, without time an timezone, there could be minor issues a couple of times each year.
If user context is not necessary here (example of sending an email with regional context), storing the datetime in UTC as @javiereguiluz mentioned earlier would be the generic way to go. APP here has the domain/business logic and should be responsible for date, time and timezone consistency.

Reply:
@weaverryan that works if you only use the value on the PHP side. But your `removeExpiredTokens` uses the datetime field in SQL, where it would use the timezone of the DB server.

The truth is that I don't know any way to write timezone-safe code with MySQL. Synchronizing the default timezone of the PHP and MySQL servers is a must (which is also why I tend to force my PHP server to use UTC as its default timezone, and then perform timezone changes in PHP for display parts of my app if necessary).

Reply:
You are not programming a calendar. Time zones should not matter for what you want to achieve here. You want to store a specific point in time for later comparison. A local time is not safe for such comparisons, especially when it comes to DST. I think, normalizing to UTC when storing/comparing dates would be the way to go here.

Reply:
> I think, normalizing to UTC when storing/comparing dates would be the way to go here.

But is this possible? All the solutions seem to involve a custom Doctrine type to force UTC, which we cannot do here "generically" (unless we added a new UTC type to Symfony itself). I would love to force UTC, but I don't see how (generically) with Doctrine.

Reply:
And if we do NOTHING (and do not convert to UTC), I believe the "only problem" would be daylight savings adjustments. Stof mentioned:

> But your removeExpiredTokens uses the datetime field in SQL, where it would use the timezone of the DB server (and so correspond to a different timestamp).

I don't think that's true, because that method converts the DateTime into a "date time string" before it makes the query. And so it is then comparing 2 date strings (the original one what was inserted and the one in the WHERE clause) that were both converted using PHP's timezone.

Reply:
I think using unix timestamps might be the best solution, as that is timezone independent IIRC.

Reply:
> But is this possible? All the solutions seem to involve a custom Doctrine type to force UTC, which we cannot do here "generically" (unless we added a new UTC type to Symfony itself). I would love to force UTC, but I don't see how (generically) with Doctrine.

Not with Doctrine as far as I know. That's why I use the custom Doctrine types 😃 

But, if you use the getter/setter defined in the trait to convert the incoming/outgoing data to a numeric Unix timestamp, I believe you're there.

Reply:
I've done that :)